Clean up, fix errors, add docstrings and automated tests.

ternus opened the pull request:
This API is great and I'm so glad it exists and has a Python library. This PR is my attempt to give back :) In this PR I:

- Used `pep8ify` to do some minor style cleanup.
- Ran `pylint` and cleaned up the warnings and errors it produced. Included .pylintrc.
- Added some basic docstrings for modules, classes, and functions. 
- Added basic integration tests. You can run them with `python -m unittest discover`.

ternus commented:
@adamhooper Hey, are you this project's maintainer? How does this look?

adamhooper commented:
On vacation. Will look next week :-)

On Tue, Aug 16, 2016, 12:24 PM Christian Ternus <notifications@github.com>
wrote:

> @adamhooper <https://github.com/adamhooper> Hey, are you this project's
> maintainer? How does this look?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/huffpostdata/python-pollster/pull/2#issuecomment-240156195>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAqPIzU_LgeH4EYa2BhlfujzEP7canKks5qgeQbgaJpZM4Jjy2_>
> .
>
-- 
Adam Hooper
http://adamhooper.com
+1-929-332-3503

ternus commented:
Oh, sorry -- have a great vacation :)

--
Christian J. Ternus | AC3T | @ternus <http://twitter.com/ternus> |
cternus.net | 617-710-8955

On Tue, Aug 16, 2016 at 12:25 PM, Adam Hooper <notifications@github.com>
wrote:

> On vacation. Will look next week :-)
>
> On Tue, Aug 16, 2016, 12:24 PM Christian Ternus <notifications@github.com>
> wrote:
>
> > @adamhooper <https://github.com/adamhooper> Hey, are you this project's
> > maintainer? How does this look?
> >
> > —
> > You are receiving this because you were mentioned.
> >
> >
> > Reply to this email directly, view it on GitHub
> > <https://github.com/huffpostdata/python-pollster/
> pull/2#issuecomment-240156195>,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/AAAqPIzU_
> LgeH4EYa2BhlfujzEP7canKks5qgeQbgaJpZM4Jjy2_>
> > .
> >
> --
> Adam Hooper
> http://adamhooper.com
> +1-929-332-3503
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/huffpostdata/python-pollster/pull/2#issuecomment-240156560>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEtcgVydcWdPICzyGN7BU-TnuWo94BFks5qgeRogaJpZM4Jjy2_>
> .
>

adamhooper commented on pollster/pollster.py:
```diff
@@ -1,41 +1,52 @@
             response = urllib2.urlopen(url)
-        except urllib2.HTTPError, e:
-            res = e.read()
+        except urllib2.HTTPError as url_exc:
```
Have you tested this code path? Please do one of the following:

* Add a unit test to test for a ValueError and another to test for a PollsterException; OR
* Undo this particular change

ternus replied:
The only change from the original is renaming the variable -- it doesn't change the underlying logic at all.

adamhooper replied:
I mean this plus the next 10 lines or so.

adamhooper commented:
Thank you for this awesome pull request! I have one concern (the urllib2.HTTPError-handling logic). Aside from that, everything looks great. If you'd like to merge the rest of it ASAP, we could move the urllib2.HTTPError change to another pull request....

ternus commented:
I've reworked the logic slightly. I have yet to encounter a situation where this logic is triggered:
 
```
                r_msg = dict(json.loads(res))
                if r_msg.has_key('errors'):
                    msg += "[%s]" % r_msg['errors'][0]
```

but assume that in some cases the backend will return a JSON response containing `errors`. (In my tests, `json.loads(res)` would occasionally fail and be of type `boolean`), hence the ValueError check.)

Anyway, I've reworked this to make sure a PollsterException is raised in all circumstances, made the `urlopen` exception more informative (including the HTTP error code and reason) and added a couple of tests for obvious failure cases.

adamhooper commented:
This is looking fantastic. I'm thrilled about the unit tests.

Your `bad_param` test actually _does_ trigger the `['errors'][0]` logic, and it does what we'd expect, so that's great.

Nit: please add a space before `[%s]`. After that, I'll merge.

ternus commented:
`pollster.pollster.PollsterException: An error occurred. URL: http://elections.huffingtonpost.com/pollster/api/polls?param=bad Reason: 400 Bad Request [param is an invalid parameter]`

Huh, so it does :) Anyway, amended the most recent commit to add the space rather than adding another commit.

adamhooper commented:
Urgh, in the future please never amend. I'll merge and if the space doesn't make it through, I'll write an extra commit myself.

Thank you! You've been wonderful.

ternus commented:
> Urgh, in the future please never amend.

Why not? It's equivalent to creating two commits and squashing them, which I've been asked to do when e.g. submitting a PR to Google open-source projects.

adamhooper commented:
First, I'll say what happened: the amend never made it into this pull request.

Now I'll explain why :).

The `amend` command has a misleading name: it doesn't actually edit a commit, because git commits are immutable -- that's fundamental to the way git works. Instead, `amend` creates a _new_ commit and hides the old one by changing the `HEAD` of its branch.

For instance, today this pull request started at `8442e89`, and you appended `412255d` to it. Then when you `amend`ed, you _removed_ `412255d` and added a new one (let's call it `b00b000` because I don't know what it actually was). The `master` on your computer ended with `8442e89` -> `b00b000`, but on GitHub, `origin/master` was `8442e89` -> `412255d`.

But you can't just _push_ the `b00b000` to `origin/master`, because `b00b000` goes on top of `8442e89` but `origin/master`'s head is `412255d`, which `b00b000` can't apply to. So you have to do a `git push --force`.

But if you do _that_, then all of us on the rest of the Internet can't _pull_ your `origin/master`: it's at `8442e89` -> `b00b000` but everybody's `HEAD` is `412255d`, so it isn't a fast-forward merge.

The workaround is for everyone else on the Internet to force-update your branch. Which defeats the purpose of git.

`git commit --amend` is useful for fixing commits on your own computer. Once you've pushed something to the Internet, though, `git commit --amend` is suddenly a high-cost operation: essentially, everybody who has seen your branch (including GitHub) has to handle it specially.

It's much simpler to just create a new commit.

ternus commented:
Thank you for the git explanation, but it really wasn't necessary. I agree that it's very bad practice to edit the history **of the upstream master branch**; that's not what's happening here.

The `--amend` was *not* why it never made it into the pull request -- it never made it in because I didn't push the amended commit:

```
diff --git a/pollster/pollster.py b/pollster/pollster.py
index a07c4d8..61cc257 100644
--- a/pollster/pollster.py
+++ b/pollster/pollster.py
@@ -46,7 +46,7 @@ class Pollster(object):
             try:
                 r_msg = dict(json.loads(res))
                 if r_msg.has_key('errors'):
-                    msg += "[%s]" % r_msg['errors'][0]
+                    msg += " [%s]" % r_msg['errors'][0]
             except ValueError:
                 pass
             raise PollsterException, msg
```

Github handles amended and squashed commits gracefully -- you can see [this recent PR](https://github.com/google/google-api-python-client/pull/261) where I squashed several commits into one, **also rewriting history**. The pull-request API recognized that this happened and dealt with it properly.

The accepted workflow for many projects I've worked on is:

 * Dev creates their own private feature branch (or, in this case, fork).
 * Dev pushes commits to their own private feature branch, which may be super messy, especially when elements such as : "Fix tests" "Fix tests for real this time" "Try number 3", etc.
 * When work is finished, dev:
    * rebases their commits on top of master with `git pull --rebase`
    * squashes their commits so each commit represents a discrete change
    * submits a pull request.
 * Project maintainer reviews the pull request, and if necessary makes suggestions.
 * Dev **amends** pull request, either adding new commits (such as the "add more tests" commit I made) or amending to e.g. fix a typo.
 * Project maintainer merges the commits from dev's feature branch.

This keeps the history clean and easy to read.

Please see http://jamescooke.info/git-to-squash-or-not-to-squash.html and https://www.atlassian.com/git/articles/git-team-workflows-merge-or-rebase/ for more discussion on this topic, but tl;dr:

![Rebase](https://pbs.twimg.com/media/BZWjU5ZIAAAMhQA.jpg)

ternus commented:
But, to be clear, I'm perfectly happy to avoid changing any history at all with respect to this project if that's your preference.

adamhooper commented:
Ooh awesome -- sorry for patronising: you obviously know your stuff. I wasn't aware GitHub supports this workflow.

I think for this project, I prefer new commits to amends. That way, the new commits show up in this GitHub conversation thread. I understand the temptation to squash commits, but merging seems to me like the maintainer's responsibility -- otherwise the maintainer needs to be vigilant to avoid errors (malicious or not) in the squash itself.

GitHub added a new feature in April that empowers maintainers who want to squash: https://github.com/blog/2141-squash-your-commits. So let's stick with that.

Thanks!

ternus commented:
Sounds good - will do. Thanks!

On Aug 22, 2016 15:28, "Adam Hooper" <notifications@github.com> wrote:

> Ooh awesome -- sorry for patronising: you obviously know your stuff. I
> wasn't aware GitHub supports this workflow.
>
> I think for this project, I prefer new commits to amends. That way, the
> new commits show up in this GitHub conversation thread. I understand the
> temptation to squash commits, but merging seems to me like the maintainer's
> responsibility -- otherwise the maintainer needs to be vigilant to avoid
> errors (malicious or not) in the squash itself.
>
> GitHub added a new feature in April that empowers maintainers who want to
> squash: https://github.com/blog/2141-squash-your-commits. So let's stick
> with that.
>
> Thanks!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/huffpostdata/python-pollster/pull/2#issuecomment-241522794>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEtcjp69OFyZoL1n6G1VkXNmE6fx7KJks5qifhdgaJpZM4Jjy2_>
> .
>
